### PR TITLE
Add List API Endpoint for `RemoteRepository` and `RemoteOrganization`

### PR DESCRIPTION
**Pull request by saadmk11**
closes https://github.com/readthedocs/readthedocs.org/issues/7509

### CODE REVIEW

**Review by humitos — COMMENTED** [review]
This is a good start. However, I'm not sure if we have to implement this endpoint before https://github.com/readthedocs/readthedocs.org/issues/2759 (and https://github.com/readthedocs/readthedocs.org/issues/7169) because we are changing how `RemoteRepository` will behave completely.

Maybe the API response won't change, but would be good to double check that, tho. Want to take a look at this?

Also, we may need to be able to filter by user account (github, gitlab, etc) as we are currently doing on API v2:

- https://readthedocs.org/api/v2/remote/repo/?org=162925 (filter by organization)
- https://readthedocs.org/api/v2/remote/repo/?own=gitlab (filter by account)

(you can see your own API URLs with the `org=` and `own=` values at https://readthedocs.org/dashboard/import/)

The idea of using `/remote` as a namespace may be good as well. I'd call it `/remotes/repositories/` in the new APIv3.

In APIv2 we also have `/remote/account` and `/remote/org` that I think they will be required here as well (could be done in a different PR, but should be considered in the whole implementation).

If you have the time and energy to work on these things, I'm happy to help you with the process, chat and/or talk about it.

**Comment by saadmk11** [comment]
> This is a good start. However, I'm not sure if we have to implement this endpoint before #2759 (and #7169) because we are changing how `RemoteRepository` will behave completely.
> 
> Maybe the API response won't change, but would be good to double check that, tho. Want to take a look at this?

Sure! 👍  I did not see the changes being done with remote repositories. I'll look into the changes and see if there is anything we need to change about this API.

> Also, we may need to be able to filter by user account (github, gitlab, etc) as we are currently doing on API v2:

I was thinking about that, but did not know what are the requirements for API V3 are. Thanks for letting me know 💯 


> The idea of using `/remote` as a namespace may be good as well. I'd call it `/remotes/repositories/` in the new APIv3.

This sound like a good idea. 👍 

> In APIv2 we also have `/remote/account` and `/remote/org` that I think they will be required here as well (could be done in a different PR, but should be considered in the whole implementation).

Okay, so we need all the API endpoints that are required for the import process.

> If you have the time and energy to work on these things, I'm happy to help you with the process, chat and/or talk about it.
Sure, I would love to work on this, recently working with different API's and liking it a lot, and RTD has one of the best designed API 💯 

It would be great if you are able to help me with the process, Let me know when you are free to talk or chat about it, maybe on slack?

**Comment by humitos** [comment]
> It would be great if you are able to help me with the process, Let me know when you are free to talk or chat about it, maybe on slack?

I'm on Slack on Europe time. Feel free to ping me there when you have time.

**Comment by saadmk11** [comment]
Marking this as blocked as me and @humitos talked, this should be implemented after #2759 (and #7169)

**Comment by humitos** [comment]
@saadmk11 this could be a good pull request to cleanup and finish now we have the new structure deployed.

File: docs/api/v3.rst [hunk 1]
```diff
@@ -1624,6 +1624,149 @@ Organization projects list
...
+                    "url": "https://github.com/organization",
+                    "vcs_provider": "github"
+                }
+            ]
+        }
+
+
+    The ``results`` in response is an array of remote organizations data.
+
+    :query string name: return remote organizations with matching name
```
Comment (humitos) [inline docs/api/v3.rst:1684]: I don't thinks this query string is required. I suppose that we will end up with an endpoint `/api/v3/remote/organizations/<slug>/` instead.

However, this `name=` attribute could be used as the `__contains` filter.

File: docs/api/v3.rst [hunk 2]
```diff
@@ -1624,6 +1624,149 @@ Organization projects list
        }
 
 
+Remote Organizations
+~~~~~~~~~~~~~~~~~~~~
+
+Remote Organizations are the importable organizations connected via
+``GitHub``, ``GitLab`` and ``BitBucket``.
```
Comment (humitos) [inline docs/api/v3.rst:1631]: These organizations are not "importable" :smile:

File: docs/api/v3.rst [hunk 3]
```diff
@@ -1624,6 +1624,149 @@ Organization projects list
...
+
+    .. sourcecode:: json
+
+        {
+            "count": 20,
+            "next": "api/v3/remote/repositories/?limit=10&offset=10",
+            "previous": null,
+            "results": [
+                {
+                    "organization": {
```
Comment (humitos) [inline docs/api/v3.rst:1732]: I think that by default we shouldn't return the `organization` object in the response. We can make this attribute expandable (`?expand=organization`).

File: docs/api/v3.rst [hunk 4]
```diff
@@ -1624,6 +1624,149 @@ Organization projects list
...
+                    "admin": true
+                }
+            ]
+        }
+
+
+    The ``results`` in response is an array of remote repositories data.
+
+    :query string name: return remote repositories with matching name
+    :query string vcs: return remote repositories for specific vcs (``git``, ``svn``, ``hg`` or ``bzr``)
```
Comment (humitos) [inline docs/api/v3.rst:1763]: I think we can remove this one. Is there a good use case?

File: docs/api/v3.rst [hunk 5]
```diff
@@ -1624,6 +1624,149 @@ Organization projects list
...
+    :query string vcs_provider: return remote repositories for specific vcs provider (``github``, ``gitlab`` or ``bitbucket``)
+    :query string organization: return remote repositories for specific remote organization (Using remote organization ``pk``)
```
Comment (humitos) [inline docs/api/v3.rst:1765]: I think we should use the remote organization `slug`, instead. Most of the API is based/using `slug` now.

File: docs/api/v3.rst [hunk 6]
```diff
@@ -1624,6 +1624,149 @@ Organization projects list
...
+                    "vcs_provider": "github",
+                    "admin": true
+                }
+            ]
+        }
+
+
+    The ``results`` in response is an array of remote repositories data.
+
+    :query string name: return remote repositories with matching name
```
Comment (humitos) [inline docs/api/v3.rst:1762]: Same here. I think we should use this as `__contains`. If this is the use case, we should explain it better.

We could also use `name__contains=` as attribute name as well. How are we calling these attributes in other APIv3 endpoints? We should follow the same pattern.

File: readthedocs/api/v3/serializers.py [hunk 7]
```diff
@@ -891,3 +892,56 @@ class Meta:
...
+        read_only_fields = fields
+
+
+class RemoteRepositorySerializer(serializers.ModelSerializer):
+    admin = serializers.SerializerMethodField('is_admin')
+    organization = RemoteOrganizationSerializer()
+
+    class Meta:
+        model = RemoteRepository
+        fields = [
```
Comment (humitos) [inline readthedocs/api/v3/serializers.py:920]: I think we can make `project` field expandable as well (`?expand=project`).

File: docs/api/v3.rst [hunk 8]
```diff
@@ -1624,6 +1624,149 @@ Organization projects list
...
+                        "created": "2019-04-29T10:00:00Z",
+                        "modified": "2019-04-29T12:00:00Z",
+                        "name": "Organization Name",
+                        "pk": 1,
+                        "slug": "organization",
+                        "url": "https://github.com/organization",
+                        "vcs_provider": "github"
+                    },
+                    "avatar_url": "https://avatars3.githubusercontent.com/u/test-rtd?v=4",
+                    "clone_url": "https://github.com/rtd/project.git",
```
Comment (humitos) [inline docs/api/v3.rst:1743]:
```suggestion
                    "clone_url": "https://github.com/organization/project.git",
```

Make sure the example matches the organization. There are some URLs pointing to `rtd` but the organization object says `organization`.

File: readthedocs/api/v3/serializers.py [hunk 9]
```diff
@@ -891,3 +892,56 @@ class Meta:
...
+            'admin',
+            'avatar_url',
+            'ssh_url',
+            'clone_url',
+            'html_url',
+            'vcs',
+            'vcs_provider',
+            'created',
+            'modified',
+            'organization',
```
Comment (humitos) [inline readthedocs/api/v3/serializers.py:934]: I think we are missing some fields here:

* `private`
* `default_branch`

File: readthedocs/api/v3/tests/test_remoteorganizations.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        )
+        social_account = fixture.get(SocialAccount, user=self.me, provider=GITHUB)
+        fixture.get(
+            RemoteOrganizationRelation,
+            remote_organization=self.remote_organization,
+            user=self.me,
+            account=social_account
+        )
+
+    def test_remote_organization_list(self):
```
Comment (humitos) [inline readthedocs/api/v3/tests/test_remoteorganizations.py:38]: Would be good to have a test that uses the `name=` filter which is the one we will use the most.

File: readthedocs/api/v3/tests/test_remoterepositories.py [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+            admin=True
+        )
+        fixture.get(
+            RemoteOrganizationRelation,
+            remote_organization=self.remote_organization,
+            user=self.me,
+            account=social_account
+        )
+
+    def test_remote_repository_list(self):
```
Comment (humitos) [inline readthedocs/api/v3/tests/test_remoterepositories.py:64]: Same here. Would be good to have a test that uses `?name=` filter.

File: readthedocs/api/v3/views.py [hunk 12]
```diff
@@ -414,3 +430,36 @@ def get_view_name(self):
...
+
+class RemoteOrganizationViewSet(APIv3Settings, ListModelMixin, GenericViewSet):
+    model = RemoteOrganization
+    serializer_class = RemoteOrganizationSerializer
+    filterset_class = RemoteOrganizationFilter
+    queryset = RemoteOrganization.objects.all()
+    permission_classes = (IsAuthenticated,)
+
+    def get_queryset(self):
+        return super().get_queryset().api(self.request.user)
```
Comment (humitos) [inline readthedocs/api/v3/views.py:465]: I'd create a `RemoteQuerySetMixin` in `api/v3/mixins.py` with this line, and use it in both views. We are doing that for the other views.

`RemoteRepositoryViewSet` may need to overwrite it tho, to add the `.annotate`, but that's fine.

File: readthedocs/api/v3/views.py [hunk 13]
```diff
@@ -414,3 +430,36 @@ def get_view_name(self):
...
+                RemoteRepositoryRelation.objects.filter(
+                    remote_repository=OuterRef('pk'),
+                    user=self.request.user,
+                    admin=True
+                )
+            )
+        )
+        return queryset.select_related('organization').order_by(
+            'organization__name', 'full_name'
+        ).distinct()
```
Comment (humitos) [inline readthedocs/api/v3/views.py:454]: `.select_related` will be only required if `?expand=organization` is passed.

**Review by humitos — CHANGES_REQUESTED** [review]
Looks great! I left some comment with a few changes I'd like to see before merging this PR.

I'm also pinging @agjohnson here who is the one using this endpoint for the search-as-you-type in the new theme.

File: docs/api/v3.rst (hunk 4 above)
Comment (saadmk11) [reply docs/api/v3.rst:1763]: I don't think so, `vcs_provider` should be enough, but I saw it in the API v2, so I added it as I did not know the specific requirements for the API v3. :)

File: docs/api/v3.rst (hunk 6 above)
Comment (saadmk11) [reply docs/api/v3.rst:1762]: We are not using anything like `__contains`, for example:

https://github.com/readthedocs/readthedocs.org/blob/901100594bf6549eda973be84b1ab7ad44379d64/readthedocs/api/v3/filters.py#L20


https://github.com/readthedocs/readthedocs.org/blob/901100594bf6549eda973be84b1ab7ad44379d64/docs/api/v3.rst#L480

Comment (humitos) [reply docs/api/v3.rst]: Gotcha!

File: docs/api/v3.rst [hunk 14]
```diff
@@ -1681,7 +1681,7 @@ Remote Organization listing
 
     The ``results`` in response is an array of remote organizations data.
 
-    :query string name: return remote organizations with matching name
+    :query string name__contains: return remote organizations with containing the name
```
Comment (humitos) [inline docs/api/v3.rst:1684]: Oh, sorry. If we are already using `name=` to filter by `icontains`, we should use the same here. No need to change it to `__contains`. I'm sorry for the confusion here.

File: docs/api/v3.rst [hunk 15]
```diff
@@ -1759,10 +1805,12 @@ Remote Repository listing
 
     The ``results`` in response is an array of remote repositories data.
 
-    :query string name: return remote repositories with matching name
-    :query string vcs: return remote repositories for specific vcs (``git``, ``svn``, ``hg`` or ``bzr``)
+    :query string name__contains: return remote repositories containing the name
     :query string vcs_provider: return remote repositories for specific vcs provider (``github``, ``gitlab`` or ``bitbucket``)
-    :query string organization: return remote repositories for specific remote organization (Using remote organization ``pk``)
+    :query string organization: return remote repositories for specific remote organization (Using remote organization ``slug``)
```
Comment (humitos) [inline docs/api/v3.rst:1810]:
```suggestion
    :query string organization: return remote repositories for specific remote organization (using remote organization ``slug``)
```

I think that should be lower case :)

File: readthedocs/api/v3/views.py [hunk 16]
```diff
@@ -449,17 +465,26 @@ def get_queryset(self):
                 )
             )
         )
-        return queryset.select_related('organization').order_by(
-            'organization__name', 'full_name'
-        ).distinct()
+
+        if is_expanded(self.request, 'organization'):
+            queryset = queryset.select_related('organization')
```
Comment (humitos) [inline readthedocs/api/v3/views.py:470]: Nice! I wasn't aware of the `is_expanded` function!

**Review by humitos — COMMENTED** [review]
This is great!

We only need to revert the `name__contains` :angel:

**Review by humitos — APPROVED** [review]
:tada:

**Comment by humitos** [comment]
I'm blocking this PR only on @agjohnson 's review.

**Comment by agjohnson** [comment]
I don't have any particular feedback here, the PR looks great. If there is anything that I need changed it can be updated later.

**Comment by humitos** [comment]
Thanks @saadmk11 for all your work here! I'm merging this PR.

**Pull request merged**